File: plugins/luabindings/tutorial/1_containers.lua
-- Test/demonstrate basic faudes containers 

-- Turn on debugging messages
-- faudes.Verbosity(0)

-- ------------------------------------------
-- Plain index set (eg StateSet)
-- ------------------------------------------

-- Announce
print("################# Set of indicees")

-- Create plain set of indices
iset = faudes.IndexSet()

-- Have a name
iset:Name("PlainIndexContainer")

-- Insert some indicees
iset:Insert(2)
iset:Insert(4)
iset:Insert(8)
iset:Insert(10)

-- Print the set via lua string conversion
print(iset)

-- Access via iterator
print("Iterating over IndexSet:",iset:Name())
iit=iset:Begin()
while iit~=iset:End() do
  -- Dereference iterator 
  print(iit:Index())
  -- Increment iterator
  iit:Inc()
end

-- Simple assignment gives a reference, not a copy
isetR = iset

-- To create a copy, use copy constructor or copy method
i2set = faudes.IndexSet(iset)
i3set = iset:Copy()

-- Manipulate the original iset ...
iset:Erase(10)

-- ... which affects isetR, but not i2set and i3set
print(i2set)
print(i3set)
print(isetR)

-- Record test case
FAUDES_TEST_DUMP("iset orig",iset)
FAUDES_TEST_DUMP("iset ref",isetR)
FAUDES_TEST_DUMP("iset copy 2",i2set)
FAUDES_TEST_DUMP("iset copy 3",i3set)


-- ------------------------------------------
-- Set of indicees with symbolic names 
-- ------------------------------------------

-- Announce
print("################# Set of symbols")

-- Create the set
nset = faudes.EventSet()

-- Have a name
nset:Name("SymbolicIndiceesContainer")

-- Insert some elements
nset:Insert("alpha")
nset:Insert("beta")

-- Print the set
print(nset)

-- Symbol lookup
idx=nset:Index("beta")
name=nset:SymbolicName(idx)
print(name, " has index ", idx)

-- Iterate
print("Iterating over EventSet",nset:Name())
nit=nset:Begin()
while nit~=nset:End() do
  -- Variants of deref
  print("Element: ", nit:Index(), nit:Name(), nset:SymbolicName(nit), nset:Str(nit))
  -- Increment
  nit:Inc()
end 

-- Erase an element
nset:Erase("beta")

-- Create a copy of the set
n2set = nset:Copy()

-- Write to file
n2set:Write("tmp_nameset.txt")

-- ------------------------------------------
-- Boolean operations on sets
-- ------------------------------------------

print("################# Boolean operations")

aset=faudes.EventSet()
bset=faudes.EventSet()
aset:FromString('<A> "a" "b" "c" </A>')
bset:FromString('<B> "c" "d" "e" </B>')

print(aset)
print(bset)

-- Union
print(aset+bset)

-- Intersection
print(aset*bset)

-- Difference
print(aset-bset)

-- Inclusion
if aset <= aset + bset then
  print("a <= a+b     ---  OK")
end

-- Equality
if aset ~= bset  then
  print("a ~= b       ---  OK")
end

-- ------------------------------------------
-- Set of events with control-system attribute
-- ------------------------------------------

-- Announce
print("################# Set of events with control-system attribute")

-- Create the set eg from the above set
eset = faudes.Alphabet(nset)

-- Have a name
eset:Name("Alphabet")

-- Insert some more elements
eset:Insert("beta")
eset:Insert("lambda")

-- Set event to be controllabel
eset:Attribute("alpha"):SetControllable()

-- Insert with an attribute prepared
cattr=faudes.AttributeCFlags()
cattr:SetControllable()
eset:Insert("mue",cattr)

-- Set attribute of element
eset:Attribute("lambda",cattr)

-- Print the set
print(eset)

-- Iterate
print("Iterating over alphabet",eset:Name())
eit=eset:Begin()
while eit~=eset:End() do
  -- Variants of deref
  print("Element: ", eit:Index(), eit:Name(), eset:Str(eit), eset:Attribute(eit))
  -- Increment
  eit:Inc()
end


-- ------------------------------------------
-- Boolean operations on sets with attributes
-- ------------------------------------------

print("################# Boolean operations incl attributes")

aset=faudes.Alphabet()
bset=faudes.Alphabet()
aset:FromString('<A> "a" +C+ "b" +o+  "c" +o+  </A>')
bset:FromString('<B> "c" +o+ "d"      "e" +CF+ </B>')
cset=faudes.Alphabet(aset);
cset:Attribute("a"):ClrControllable()

print(aset)
print(bset)
print(cset)

-- The operators *, + and - return plain event sets
print("Operators swallow attributes:")
print(aset + bset)
print(aset - bset)
print(aset * bset)

-- Set comaprison operators (attributes ignored)
sle1 = (aset <= cset)
if sle1 then
  print("a <= c --- OK")
end
seq1 = (aset == cset)
if seq1 then
  print("a == c --- OK")
end

-- Set comparison methods (when types don't match)
sle2 = faudes.AlphabetInclusion(aset, aset + bset)               
if sle2 then
  print("a <= a + b --- OK")
end
seq2 = faudes.AlphabetEquality(aset - bset + aset,aset)
if seq2 then
  print("a - b + a == a --- OK")
end


-- Technical detail. By std. Lua semantics, the order operators "==", "<=" 
-- and "<" do not allow for overloading with mixed types. E.g., no matter 
-- what comparison function we implement, "==" is meant to insist in a 
-- matching type. It is only due to a rather crude hack that the below
-- code fragment evaluates as the one may expect. We recommend that you
-- avoid the issue alltogether and use the above functions.

sle3 = (aset <= aset + bset)               -- avoid this: types dont match
if sle3 then
  print("a <= a + b --- OK")
end
seq3 = (aset == aset - bset + aset )       -- avoid this: types dont match
if seq3 then
  print("a == a - b + a --- OK")
end


-- Union, maintain attributes
union=faudes.Alphabet()
faudes.AlphabetUnion(aset,bset,union)
print("AlphabetUnion(...) maintains attributes");
print(union)

-- Intersection, maintain attributes
intersection=faudes.Alphabet()
faudes.AlphabetIntersection(aset,bset,intersection)
print("AlphabetIntersection(...) maintains attributes");
print(intersection)

-- Difference, maintain attributes
difference=faudes.Alphabet()
faudes.AlphabetDifference(aset,bset,difference)
print("AlphabetDifference(...) maintains attributes");
print(difference)

-- Record test case
FAUDES_TEST_DUMP("union with attrib.",union)
FAUDES_TEST_DUMP("intersection with attrib.",intersection)
FAUDES_TEST_DUMP("difference with attrib.",difference)
FAUDES_TEST_DUMP("inclusion with attrib.",sle1)
FAUDES_TEST_DUMP("equality with attrib.",seq1)
FAUDES_TEST_DUMP("inclusion with attrib.",sle2)
FAUDES_TEST_DUMP("equality with attrib.",seq2)
FAUDES_TEST_DUMP("inclusion with attrib.",sle3)
FAUDES_TEST_DUMP("equality with attrib.",seq3)

-- ------------------------------------------
-- Transitions relations
-- ------------------------------------------

-- Announce
print("################# Set of transitions")

-- Create the set
tset = faudes.TransSet()

-- Have a name
tset:Name("TransitionRelation")

-- Insert some transitions: by component
tset:Insert(1,"alpha",2)
tset:Insert(2,"beta",1)
tset:Insert(2,"mue",3)

-- Insert some transitions: prepared
trans = faudes.Transition(3,"lambda",1)
tset:Insert(trans)

-- Print the set
tset:Write()

-- Iterate
print("Iterating over transition set",tset:Name())
tit=tset:Begin()
while tit~=tset:End() do
  -- Variants of deref
  print("Element: ", tit:Transition(), tset:Str(tit))
  -- Increment
  tit:Inc()
end


-- Sort in X2-Ev-X1 order
trset = faudes.TransSetX2EvX1()
trset:Name("ReverseTransitionSet")
tset:ReSort(trset)

-- Print the set
print("Transitionset in reverse order")
trset:Write()


-- ------------------------------------------
-- Vectors
-- ------------------------------------------

-- Announce
print("################# Vector of alphabets")

-- Create a vector of alphabets 
-- Note: corefaudes also provides vectors of event sets, generators and systems
avec = faudes.AlphabetVector()

-- Append alphabet to the vector (takes a copy)
avec:PushBack(aset)
avec:PushBack(bset)

-- Access entries (provides reference)
avec[0]:Insert("alpha")
avec[0]:Attribute("alpha"):SetControllable()

-- Access entries (yes, the [] operator does provide a reference)
aset=avec[0]
aset:Insert("beta")

-- Access entries (fine for function arguments, too)
faudes.AlphabetUnion(avec[0], avec[1], avec[2])
faudes.AlphabetIntersection(avec[0], avec[1], avec[3])

-- Set entries assignment (takes a copy)
avec[4]=aset
avec[5]=avec[2]

-- Explicit resize (here: extend)
avec:Size(8)

-- Report
avec:Write()

-- Record test case
FAUDES_TEST_DUMP("vect test",avec)

-- Validate
FAUDES_TEST_DIFF()

File: plugins/pybindings/tutorial/1_containers.py
## Test/demonstrate basic faudes containers 

## import our module (lazy variant, make global)
from faudes import *

## Set level of verbosity
## Verbosity(1)

## ##########################################
## Set of indicees (eg used as StateSet)
## ##########################################

## Announce
print("################# Set of indicees")

## Instantiate set of indices
## (relevant faudes typed objects all support default/copy/file constructors)
i1 = IndexSet()                     # default constructor (empty set)
i2 = IndexSet(i1)                   # copy constructor
i3 = IndexSet("data/1_states.fts")  # construct from file (faudes token stream)

## Edit programatically
i1.Name("Indicees")                 # give a name (cosmetic)
i1.Insert(2)                        # insert element
i1.Insert(4)                        # insert element
i1.Insert(10)                       # insert element
i1.Erase(2)                         # remove element

## Print the set for inspection 
print(f'A small IndexSet with two elements: {i1}')

## Read from string (faudes token streanm format)
i1.FromString('<Q> 2 4 8 10 </Q>')
print(f'An IndexSet from a FAUDES token stream: {i1}')

## Read from Python list
i2.FromList([1, 2, 3, 4])
print(f'An IndexSet from a Python list: {i2}')

## Iterate, Python style
print(f'Iterating over the IndexSet "{i1.Name()}" (Python style)')
for q in i1:
  print(q)

## Convert back to Python list
i1l = list(i1)
print(f'Obtaining a Python list from an IndexSet: {i1l}')

## Assignment
i1R = i1           # gives a reference, ***not*** a copy
i2 = IndexSet(i1)  # use the copy contructor for an actual copy
i3 = i1.Copy()     # use the copy method for an actual copy
i1R.Name("Q'")

## Manipulate the original set ...
print(f'Assignment vs. Copy:  erase one item affects the reference, not the copy')
i1.Erase(10)
print(f'Assignment vs. Copy:  orig {i1} ref {i1R} copy {i2}')

## All faudes sets support STL style iteration (legacy)
print(f'Iterating over the IndexSet "{i1.Name()}" (STL style)')
it=i1.Begin()
while it!=i1.End():
  print(it.Index())
  it.Inc()

## Record test case
TestDump("iset orig",i1)
TestDump("iset ref",i1R)
TestDump("iset copy 2",i2)
TestDump("iset copy 3",i3)


## ##########################################
## Set of indicees with symbolic names 
## ##########################################

## Announce
print("################# Set of symbolic names")

## Create a NameSet (alias EventSet)
n1 = NameSet()

## Edit programmatically
n1.Name("SetOfSymbols")   # give it a name
n1.Insert("alpha")        # insert a symbol
n1.Insert("beta")         # insert a symbol
n1.Insert("lambda")       # insert a symbol
n1.Insert("mue")          # insert a symbol
n1.Erase("lambda")        # remove a symbol

## Print a set
print(f'A NameSet: {n1}')

## Set from string (using static method as constructor, since "n2" does not yet exist))
n2=NameSet.NewFromString('<S> "alpha" "beta" "mue" "lambda" </S>')
print(f'A NameSet from a FAUDES token stream: {n2}')

## Set from Python list
n3=NameSet.NewFromList(["wellcome", "to", "libFAUDES"])
print(f'A NameSet from a Python list: {n3}')

## Symbol lookup
bidx=n1.Index("beta")
bname=n1.SymbolicName(bidx)
print(f'"{bname}" has index {bidx}')

## Iterate Python style (STL style is also available; see above IndexSet example))
print(f'Iterating over NameSet "{n1.Name()}" (Python style)')
for e in n1: 
  print(f'Element: idx {e} symbol "{n1.SymbolicName(e)}" info "{n1.Str(e)}"')

## Erase an element by symbolic name or index
n1.Erase("mue")
n1.Erase(bidx)

## Assignment
n1R = n1        # have a reference (not a copy)
n2  = n1.Copy() # have a copy

## Write to file
n2.Write("tmp_nameset.fts")


## ##########################################
## Boolean operations on sets
## ##########################################

print("################# Boolean operations")

aset=NameSet()
bset=NameSet()
aset.FromString('<A> "a" "b" "c" </A>')
bset.FromString('<B> "c" "d" "e" </B>')
print(f'Two sets A {aset} and B {bset}')

## Apply boolean operations
print(aset+bset)    # union
print(aset*bset)    # interscetion
print(aset-bset)    # difference

## Inclusion
if aset <= aset + bset:
  print("A <= A+B       ##  OK")
else:
  print("not (A <= A+B) ##  FAIL")

## Equality
if aset == bset:
  print("A == B         ##  FAIL")
else:
  print("A != B         ##  OK")


## ##########################################
## Set of events with control-system attributes
## ##########################################

## Announce
print("################# Set of events with control-system attributes")

## Create the Alphabet (eg from the an NameSet)
eset = Alphabet(n1)

## Edit
eset.Name("Sigma")   # have a name
eset.Insert("beta")     # insert more events
eset.Insert("lambda")   # insert more events

## Set existing event to be controllabel
eset.Attribute("alpha").SetControllable()

## Insert new event with attribute prepared
cattr=AttributeCFlags()
cattr.SetControllable()
eset.Insert("mue",cattr)

## Assign attribute to element
eset.Attribute("lambda",cattr)

## Report
print(f'Alphabet: {eset}')

## Iterate (Python style; STL style also available, see above IndexSet)
print("Iterating over alphabet",eset.Name())
for e in eset:
  print(f'idx {e} symbol "{eset.SymbolicName(e)}" attrib {eset.Attribute(e)}')


## ##########################################
## Boolean operations on sets with attributes
## ##########################################

print("################# Boolean operations incl attributes")

aset=Alphabet()
bset=Alphabet()
aset.FromString('<A> "a" +C+ "b" +o+  "c" +o+  </A>')
bset.FromString('<B> "c" +o+ "d"      "e" +C+ </B>')
cset=aset.Copy()
cset.Name('C')
cset.Attribute("a").ClrControllable()

print(f'Set A: {aset}')
print(f'Set B: {bset}')
print(f'Set C: {cset}')

## The operators *, + and - return plain event sets
print("Operators swallow attributes:")
print(f'A + B: {aset + bset}')
print(f'A - B: {aset - bset}')
print(f'A * B: {aset * bset}')

## Set comparison operators ignore attributes
sle1 = (aset <= cset)
if sle1:
  print("A <= C ## OK")
seq1 = (aset == cset)
if seq1:
  print("A == C ## OK")

## Set comparison methods (when types don't match)
sle2 = AlphabetInclusion(aset, aset + bset)               
if sle2:
  print("A <= A + B ## OK (RHS is a plain NameSet)")
seq2 = AlphabetEquality(aset - bset + aset,aset)
if seq2:
  print("A - B + A == A ## OK (LHS is a plain NameSet)")


## Union, maintain attributes where possible
union=Alphabet()
AlphabetUnion(aset,bset,union)
print("AlphabetUnion(...) maintains attributes");
print(union)

## Intersection, maintain attributes
intersection=Alphabet()
AlphabetIntersection(aset,bset,intersection)
print("AlphabetIntersection(...) maintains attributes");
print(intersection)

## Difference, maintain attributes
difference=Alphabet()
AlphabetDifference(aset,bset,difference)
print("AlphabetDifference(...) maintains attributes");
print(difference)

## Record test case
TestDump("union with attrib.",union)
TestDump("intersection with attrib.",intersection)
TestDump("difference with attrib.",difference)
TestDump("inclusion with attrib.",sle1)
TestDump("equality with attrib.",seq1)
TestDump("inclusion with attrib.",sle2)
TestDump("equality with attrib.",seq2)

## ##########################################
## Transitions relations
## ##########################################

## Announce
print("################# Set of transitions")

## Create the set
t1 = TransSet()

## Have a name
t1.Name("Delta")

## Insert some transitions: by component
t1.Insert(1,"alpha",2)
t1.Insert(2,"beta",1)
t1.Insert(2,"mue",3)

## Insert some transitions: by transition object
trans = Transition(3,"lambda",1)
t1.Insert(trans)

## Print the set
print(f'TransSet {t1}')

## Read transitions from string (no symbolic names)
t2=TransSet.NewFromString("""<T2>
  1 2 3
  4 5 6
</T2>""")
print(f'TransSet from string: {t2}')

## Read transitions from Python list (no symbolic names)
t3=TransSet.NewFromList([
  ( 7, 8, 9),
  (10,11,12)
])
t3.Name('T3')
print(f'TransSet from Python list: {t3}')

## Obtain a Python list
t2l=list(t2)
print(f'TransSet T2 as Python list: {t2l}')

## Iterate (Python Style)
print(f'Iterating over transition set "{t1.Name()}"')
for t in t1:
  print(f'Transition: x1={t.X1} ev={t.Ev} x1={t.X2}')

## Retrieve set of all states and inspect on per state basis
print(f'Inspecting transition set "{t1.Name()}" on a per state basis')
xset=t1.States()
for x1 in xset:
  print(f'State X1 {x1}: active events {t1.ActiveEvents(x1)} successors {t1.SuccessorStates(x1)}')

## Sort in X2-Ev-X1 order
t1rev = TransSetX2EvX1()
t1rev.Name("Delta_Reverse")
t1.ReSort(t1rev)

## Inspect reverse transition relation on per state bais
print(f'Inspecting transition set "{t1rev.Name()}" on a per state basis')
for x2 in xset:
  print(f'State X2 {x2}: incomming events {t1rev.IncommingEvents(x2)} predecessors {t1rev.PredecessorStates(x2)}')


## ##########################################
## Vectors
## ##########################################

## Announce
print("################# Vector of alphabets")

## Create a vector of alphabets 
## (libFAUDES provides vectors of event sets, generators and systems with similarsyntax/semantics)
avec = AlphabetVector()

## Append alphabets to the vector (takes a copy, the vector owns its elements)
avec.PushBack(aset)
avec.PushBack(bset)

## Access entries (the operator [] provides an editable reference)
avec[0].Insert("alpha")
avec[0].Attribute("alpha").SetControllable()

## Access entries (yes, the [] operator does provide a reference)
aset=avec[0]
aset.Insert("beta")

## Access entries (fine for function arguments, too)
AlphabetUnion(avec[0], avec[1], avec[2])
AlphabetIntersection(avec[0], avec[1], avec[3])

## Set entries assignment (takes another copy, the vector owns its data)
avec[4]=aset
avec[5]=avec[2]

## Explicit resize (here: extend)
avec.Size(8)

## Report
avec.Write()

## Record test case
TestDump("vect test",avec)

## Validate
TestDiff()
